### DISCUSSION TITLE
topo_mask logic in point-stat

### DISCUSSION THREAD

**Discussion by fearon10** [dtcenter/METplus#2232] [Configuration] [answered]
Hi There,

I'm running point-stat verifying model grid points against RAOBs. I'm not entirely sure I have the topo_mask logic correct, as I don't observations being rejected based on topography or at a minimum I don't see topo checking being performed. I have the verbosity set at v5. My topo_mask setting is the following in my point-stat config

topo_mask = {
    flag = TRUE;
    FILE_NAME = ["PYTHON_NUMPY"];
    field = { name = "/read_neptune.py -f ci_quarterdeg_E0096P4L105_20210806T12_PT000000H00M00S.hdf5 -m NEPTUNE -l '0' -lt 'model_terrain' -v 'terrain'";}
    regrid = { method = BILIN; width=2;}
    use_obs_thresh = ge-100&&le100;
    interp_fcst_thresh = ge-50&&le50;
}

I also have following in my config for message_type and message_type_group_map, respectively

message_type   = ["RAOB"];

message_type_group_map = [
   { key = "SURFACE"; val = "ADPSFC,SFCSHP,MSONET,RAOB";          },

I noted the "use_obs_thresh" filtering goes by topo - station elevation. Based on that, I was using "use_obs_thresh = le0; , but that choice showed no change. I've attached my log file in case it's helpful. 

Thank you for the assistance. 
Matt

[log.point_stat.txt](https://github.com/dtcenter/METplus/files/11931120/log.point_stat.txt)

**Comment by DanielAdriaansen** [comment]
Hi @fearon10. Thank you for providing your log file and information about your configuration. I notice in the log file output that you sent that the level you are evaluating is 500 hPa:
```
DEBUG 2: Processing air_temperature/500.0 versus air_temperature/P50000, for observation type RAOB, over region FULL, for interpolation method NEAREST(1), using 584 matched pairs.
```
Could you try selecting a lower level and see if any points are filtered based on the topography?

**Reply by fearon10** [reply]
Hi @DanielAdriaansen. Thanks for the quick reply. Great suggestion on trying a lower level. I tried 950, 900, and 850 hPa with the same results, no topography filtering. Maybe my use_obs_thresh logic is not correct? Thanks again.

**Comment by j-opatz** [comment]
Hi Matt,

I'm sorry to hear you're having trouble with the topo_mask dictionary logic. I reviewed your `use_obs_thresh` and it appears fine, and you were correct in your assumption that changing `use_obs_thresh` to `le0.0` should have dropped a few of your observation points (unless your RAOB data just happens to be at the exact same height as the topo field!).

I think I might have to test around your set up to find out what's wrong. Would you be able to send over the files necessary to run 1 example of where you're using this set up? You can use our [FTP instructions to send data over](https://github.com/dtcenter/METplus/discussions/954), or you can point us to the NOAA system (Jet, Hera, etc) and we should be able to pull a copy from there.

**Comment by fearon10** [comment]
Hi There,

Really appreciate the assistance. I have uploaded my set up files for one example to fearon_data/ on the ftp. The files include:

obs file: out_raob_20210806T1200Z.nc4_2021080612
fcst file: i_quarterdeg_E0096P4L105_20210806T12_PT000000H00M00S.hdf5
config file: PointStatConfig_IODA_suite_NEP_myn_tie_tho_s7_2021080612_000000
climo file: cmean_1d.19790806
python script to read the fcst file: read_neptune.py
my command syntax to run point-stat: run_command.txt

This syntax works without issue except for the topo flag (as shown in the config file) which yields no ob rejection. I would think at least some obs would be rejected based on my threshold choices. I have even tested adding 3000 meters to each terrain grid cell value to force ob rejection, however, the results remain unchanged. Overall, it seems as though the topo flag is being ignored. 

Thanks,
Matt

**Reply by j-opatz** [reply]
Thanks for everything you've supplied Matt,

It looks like there may be an issue with MET calling your script. When I run the MET configuration file as is (only adjusting output directories, temporary directories, etc.), I get the following error during the topography field run:

```
 File "read_neptune.py", line 95, in <module>
    units     = ds.units
NameError: name 'ds' is not defined
```

That perplexed me, since 1) your script is called for both the forecast and topography mask and 2) you didn't have any issues in your run with the Python script.

I confirmed that the forecast field is being passed correctly and your script works there, but when it gets to the topography I can't understand what's causing the current issue. A bit of testing shows that none of your `elifs` or `if` statements are catching the `inlevtype` you're passing, even though a print statement shows it's 'model_terrain' (which is clearly your last `elif` statement). I put a catch-all `else` statement at the end to see if I could force it to take the same steps as that last `elif` but got the error:

```
  File "read_neptune.py", line 90, in <module>
    data     = ds[:]
TypeError: 'NoneType' object is not subscriptable
```

So even if I force it to go find the `terrain` variable, it seems like the script currently returns with a NoneType.

I've further confirmed this is NOT an issue with your script as a direct call to your Python script with the exact same command you're using in MET

```
python read_neptune.py -f ci_quarterdeg_E0096P4L105_20210806T12_PT000000H00M00S.hdf5 -m NEPTUNE -l '0' -lt 'model_terrain' -v 'terrain'
```

runs perfectly fine.

I'm going to tag @DanielAdriaansen and @JohnHalleyGotway on this and hopefully they can help me pinpoint where MET is missing the hand-off on this.

**Reply by j-opatz** [reply]
@fearon10 we've been working on this issue and while the initial hurdle of your script not working for the topography dictionary was solved (@DanielAdriaansen found the single quotes in your line 
`field = { name = "read_neptune.py -f ci_quarterdeg_E0096P4L105_20210806T12_PT000000H00M00S.hdf5 -m NEPTUNE -l '0' -lt 'model_terrain' -v 'terrain'";}` 
around model_terrain, terrain, and 0 were not necessary and caused a string mismatch in your script), we're still a bit at a loss as to how to how to make the topography work as you're intending. 

But we're hoping that can be solved with a little more context from you. We're having a hard time understanding the need for a topographic mask against RAOB data. With Radiosondes always launching from the surface, you shouldn't need a filter for the ground since it's always the ground (and partially why a topo_mask will allow all RAOB data through the threshold). Is this requirement coming from the desire to filter RAOBs against your model data being at odd heights? Or some other requirement? Any information on why your setup is using the topo_mask option would be useful.

**Comment by fearon10** [comment]
Thanks for the great feedback @j-opatz and @DanielAdriaansen.

I guess my ultimate question is: I am curious if my usage and syntax with the topo_mask is correct? 
Regarding RAOBs, maybe in this case the model terrain is below the station elevation and the topo_mask is working as intended? However, what if the model grid is coarser and smoother, then the model terrain may be higher than the RAOB station elevation - perhaps a rare case. I did try a case where I added 3000 meters to all model terrain grid points (ahead of read-in into MET) and the topo_mask remained inactive (i.e., no obs were rejected).

Thanks for your continued assistance.

**Reply by DanielAdriaansen** [reply, marked as answer]
Hi @fearon10 that is a great test you performed to add height to the terrain and prove you're masking is working.

There is one additional issue we uncovered that we have a question about and I think just forgot to mention in our last reply (sorry about that). In your observation RAOB file:
`out_raob_20210806T1200Z.nc4_2021080612`, there is no valid observation height information for MET to use to compare with the model terrain.

Using Python to inspect that file (which based on the header appears to have come from `ioda2nc`, is that correct?), I see the following for the `obs_hgt` variable, which is the variable that MET would be using to compare from RAOB to the model terrain:
```
>>> ds.obs_lvl.values.max()
103443.64
>>> ds.obs_lvl.values.min()
230.0
>>> ds.obs_hgt.values.max()
nan
>>> ds.obs_hgt.values.min()
nan
```
It appears the minimum and maximum `lvl` have valid data (pressure), but `hgt` does not. Can you see if you can create an observation file that has the height of each RAOB observation in meters, and try to see if your test adding 3000m to model terrain results in filtering?

**Reply by fearon10** [reply]
Hi @DanielAdriaansen 

Apologies, I did not realize the obs_hgt variable was completely missing which it looks to be. The file was created with ioda2nc, however, the obs_hgt is absent in the original ioda file as well. I'll check in with some folks here and find out if hgt is truly missing or just not included in the original ioda file. Perhaps having no height variable is the main issue. Thanks so much. Will report back soon.

**Reply by fearon10** [reply]
Hi @DanielAdriaansen and @j-opatz
Just wanted to follow up. As Daniel suggested in our last correspondence, I believe my primary issue with the topo_mask not rejecting obs is related to the absence of the hgt variable in the RAOB IODA data file. In speaking with folks in our research group, the next step is to add the hgt variable from the raw RAOB file. However, that will take some time. Otherwise, I guess I have no further questions at this time. Really appreciate the assistance. 
Thank you,
Matt
